hips/hip-9999.md

```diff
@@ -0,0 +1,483 @@
+
+Global:
+
+- `sourceURL` is added to encourage best security practices. See [Security implications](#Security-implications).
+- `apiVersion` added to allow us to update the overall plugins system API, similar to Chart API version. This will allow us to move to a new major version of Extism or other Wasm tooling independent of Helm major version.
+- `type` is added to specify the version of the API for the desired plugin type. This API defines the message scheme the plugin expects to handle, and versioning allows that to be updated within the same MAJOR version of Helm. The proposed format is `plugins.helm.sh/TYPE/VERSION`.
+
+```yaml
+name: my-custom-render-plugin # same as Helm 3
+version: 0.1.0 # same as Helm 3
+apiVersion: v1 # NEW: overall plugin system API version
+type: plugins.helm.sh/render/v1 # NEW: v1 of the "render" plugin type
+sourceURL: https://github.com/example/my-plugin # NEW
+```
+
+### Helm plugin commands
+
+- `helm plugin list` command will be updated to also show the type of plugin, whether or not it is core or custom, the plugin source, and signing key if any. Example output:
+    ```console
+    NAME        VERSION     TYPE            KIND    
+    oci         4.0.1       download        core
+    http/s      4.0.0       download        core
+    s3          1.2.3       download        custom
+    gotemplate  4.0.0       render          core
+    yamlscript  4.0.0       render          core
+    cue         0.2.0       render          custom
+    kustomize   4.0.0       post-render     core
+    my-bash     0.1.0       post-render     custom
+
+    -o wide
+
+    SIGNED      SOURCE
+    208DD36E    github.com/helm/helm
+    208DD36E    github.com/helm/helm
+    862HB19B    gitlab.com/someone/s3-helm-plugin
+    208DD36E    github.com/helm/helm
+    208DD36E    github.com/helm/helm
+    774UR67F    github.com/someone/cue-helm-plugin
+    208DD36E    github.com/helm/helm
+    no          local-dev
+    ```
+- `helm plugin install` will now install the latest release from a VCS repo URL, if a release version is not specified with `--version`
+- `helm plugin update` will check for updates, then follow the same behavior of `plugin install`
+- `helm plugin sign` will use Helm 4's built-in provenance signing mechanisms as a convenience to sign a plugin
+- `helm plugin verify` will use Helm 4's built-in provenance mechanisms as a conveniencce to verify a signed plugin
+- `helm plugin create` will add create a boilerplate plugin scaffolding, specified by `--type`. `--help` will list the types of plugins supported by your installed version of Helm
+
+Additionally commands will be added to allow packaging Chart-defined plugins within the chart (see [Distribution](#Distribution) for reasoning):
+
+- `helm chart plugin build` similar to `helm dependency build` but for plugins
+- `helm chart plugin update` similar to `helm dependency update` but for plugins
+
+### Distribution
```

Comment:
Why wouldn't we support the downloader plugins as the means to download WASM? That would provide http(s) and OCI out of the box. It would make it extensible in the same way everything else is.

Reply:
Absolutely want this. It is a goal, if we can implement that in time. Otherwise, we can start with the current options and do this as a follow up.